framework/framework.go

```diff
@@ -1,54 +1,134 @@
 	// Dependencies.
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
 	}
+	if len(config.Resources) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "config.Resources must not be empty")
+	}
 
 	newFramework := &Framework{
 		// Dependencies.
-		logger: config.Logger,
+		logger:    config.Logger,
+		resources: config.Resources,
+
+		// Internals.
+		mutex: sync.Mutex{},
 	}
 
 	return newFramework, nil
 }
 
-type Framework struct {
-	// Dependencies.
-	logger micrologger.Logger
+// AddFunc executes the framework's ProcessCreate and ProcessUpdate functions,
+// in this order. This guarantees resource creation is always done before
+// resource updates.
+func (f *Framework) AddFunc(obj interface{}) {
+	// We lock the AddFunc/DeleteFunc to make sure only one AddFunc/DeleteFunc is
+	// executed at a time. AddFunc/DeleteFunc is not thread safe. This is
+	// important because the source of truth for the kvm-operator are Kubernetes
+	// resources. In case we would run the operator logic in parallel, we would
+	// run into race conditions.
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	f.logger.Log("debug", "executing the operator's create function")
+
+	err := ProcessCreate(obj, f.resources)
+	if err != nil {
+		f.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+		return
+	}
+
+	f.logger.Log("debug", "executed the operator's create function")
+
+	f.logger.Log("debug", "executing the operator's update function")
+
+	err = ProcessUpdate(obj, f.resources)
+	if err != nil {
+		f.logger.Log("error", fmt.Sprintf("%#v", err), "event", "update")
+		return
+	}
+
+	f.logger.Log("debug", "executed the operator's update function")
+}
+
+// DeleteFunc executes the framework's ProcessDelete function.
+func (f *Framework) DeleteFunc(obj interface{}) {
+	// We lock the AddFunc/DeleteFunc to make sure only one AddFunc/DeleteFunc is
+	// executed at a time. AddFunc/DeleteFunc is not thread safe. This is
+	// important because the source of truth for the kvm-operator are Kubernetes
+	// resources. In case we would run the operator logic in parallel, we would
+	// run into race conditions.
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	f.logger.Log("debug", "executing the operator's delete function")
+
+	err := ProcessDelete(obj, f.resources)
+	if err != nil {
+		f.logger.Log("error", fmt.Sprintf("%#v", err), "event", "delete")
+		return
+	}
+
+	f.logger.Log("debug", "executed the operator's delete function")
+}
+
+// NewCacheResourceEventHandler returns the framework's event handler for the
+// k8s client's cache informer implementation. The event handler has functions
+// registered for the k8s client's add, delete and update events.
+func (f *Framework) NewCacheResourceEventHandler() *cache.ResourceEventHandlerFuncs {
+	newHandler := &cache.ResourceEventHandlerFuncs{
+		AddFunc:    f.AddFunc,
+		DeleteFunc: f.DeleteFunc,
+		UpdateFunc: f.UpdateFunc,
+	}
+
+	return newHandler
+}
```

Comment:
The framework now provides a convenient way to get a resource event handler how it is commonly used in our operators. This is a drop in for operators which makes the operator code itself way nicer. See it in action in https://github.com/giantswarm/kvm-operator/pull/177.

Reply:
Yes I like how this simplifies the operator code.